easypy/lockstep.py

```diff
@@ -0,0 +1,86 @@
+from contextlib import contextmanager
+from functools import wraps
+
+from .exceptions import TException
+
+
+class LockstepSyncMismatch(TException):
+    template = "Expected {lockstep_name} to be {expected_step}, but it is {actual_step}"
+
+
+class _LockstepInvocation(object):
+    def __init__(self, name, generator):
+        self._name = name
+        self._generator = generator
+        self._current_step = 'just-started'
+
+    def progress(self, step):
+        """Progress to the next step, confirming it's the specified one"""
+
+        try:
+            self._current_step, value = self._next_step_and_value()
+        except StopIteration:
+            raise LockstepSyncMismatch(lockstep_name=self._name,
+                                       expected_step=step,
+                                       actual_step='finished')
+        if self._current_step != step:
+            raise LockstepSyncMismatch(lockstep_name=self._name,
+                                       expected_step=step,
+                                       actual_step=self._current_step)
+        return value
+
+    def exhaust(self):
+        """Skip the remaining steps"""
+
+        for self._current_step in self._generator:
+            pass
+
+    def _next_step_and_value(self):
+        yield_result = next(self._generator)
+        if isinstance(yield_result, tuple) and len(yield_result) == 2:
+            return yield_result
+        else:
+            return yield_result, None
+
+    def __str__(self):
+        return '%s<%s>' % (self._name, self._current_step)
+
+
+def lockstep(generator_func):
+    """
+    Synchronize a coroutine that runs a process step-by-step.
+
+    Decorate a generator that yields step names to create a context manager.
+    The context object has a `.progress` method that must be called, in order,
+    with all expected step names, to make the generator progress to each step.
+    """
+
+    @contextmanager
+    def cm(*args, **kwargs):
+        invocation = _LockstepInvocation(generator_func.__name__, generator_func(*args, **kwargs))
+
+        yield invocation
+
+        try:
+            step_not_taken, _ = invocation._next_step_and_value()
+        except StopIteration:
+            # all is well - all steps were exhausted
+            invocation._current_step = 'finished'
+        else:
+            raise LockstepSyncMismatch(lockstep_name=invocation._name,
+                                       expected_step='finished',
+                                       actual_step=step_not_taken)
+
+    @wraps(generator_func)
+    def wrapper(*args, **kwargs):
+        context = cm(*args, **kwargs)
+
+        def execute():
```

Comment:
@koreno: This is a bit hacky, but I had to do it to implement the `execute` method as an attribute of the context object. I'm not sure we'll use it though - I think it would be better to create wrapper methods:

```python
@lockstep
def clusterize_lockstep(self, ...):
    # ...

def clusterize(self, ...):
    with self.clusterize_lockstep(...) as process:
        process.exhaust()
```

I put this feature is in a separate commit so that we can easily remove it if we want.